app/views/settings/dex_connector_oidcs/_form.html.slim

```diff
@@ -0,0 +1,73 @@
+= form_for [:settings, @data_holder], html: { class: "dex-connectors-form", autocomplete: "false" } do |f|
+  / Chrome may autocomplete username into text field before a password field, regardless of name
+  / This *should* prevent that (in combination with 'autocomplete=false' on the form_for; "off" no longer works)
+  / Explanation: Chrome looks for the first password field, while newer versions prefer one named "password"
+  / Firefox behaves similarly.  Hopefully they don't figure out the hidden workaround soon...
+  = f.text_field     :username, style: "display:none", value:"autocomplete trap"
+  = f.password_field :password, style: "display:none", value:"autocomplete bait"
+
+
+  .form-group class="#{error_class_for(@data_holder, :name)}"
+    = f.label :name
+    = f.text_field :name, id: "oidc_name", class: "form-control", value: @data_holder.name, required: true
+    = error_messages_for(@data_holder, :name)
+    small.form-text.text-muted
+      | Name shown to user when selecting a connector
+
+  hr
+  h4 Provider
+
+  .form-group class="#{error_class_for(@data_holder, :provider_url)}"
+    = f.label :provider_url, "Provider URL"
+    = f.text_field :provider_url, id: "oidc_provider", class: "form-control", value: @data_holder.provider_url, required: true
+    = error_messages_for(@data_holder, :provider_url)
+    small.form-text.text-muted
+      | Issuer used by OIDC Provider
+    
+  form-group class="#{error_class_for(@data_holder, :basic_auth)}"
+    = f.label :basic_auth, "Basic Auth"
+    br
+    .btn-group.btn-group-toggle data-toggle="buttons"
+      = f.label :basic_auth, nil, class: "btn btn-default #{'btn-primary active' if @data_holder.basic_auth}"
+        = f.radio_button :basic_auth, "true", checked: @data_holder.basic_auth
+        | On
+      = f.label :basic_auth, nil, class: "btn btn-default #{'btn-primary active' unless @data_holder.basic_auth}"
+        = f.radio_button :basic_auth, "false", checked: !@data_holder.basic_auth
+        | Off
+    = error_messages_for(@data_holder, :basic_auth)
+    small.form-text.text-muted
+      | Set to "Off" only if this OIDC provider requires credentials to be sent using GET (this is uncommon)
+    
+  hr
+  h4 Client
+  
+  / The 'https' and '3200' seem to be hard-coded elsewhere, so I assume they're not discoverable :/
+  - external_api_url = "https://#{Pillar.value(pillar: :apiserver)}:32000/callback"
+  .form-group class="#{error_class_for(@data_holder, :callback_url)}"
+    = f.label :callback_url, "Callback URL:"
+    br
+    form-text
+      = external_api_url
+    br
+    small.form-text.text-muted
+      | URL which will be provided to OIDC provider as the callback location
+    = f.hidden_field :callback_url, id: "oidc_callback_url", class: "form-control", value: external_api_url, readonly: true
```

Comment:
Instead of showing the url as plain text, I would simply show it as a read-only text input. If we go with that, you can remove the lines from 48 to 51.

Reply:
I initially went that way, but changed my mind. :)  My reasoning for changing to a hidden field and showing it as plain text was that a disabled text field conveys the impression that the data might vary, but isn't directly mutable by the user.  In this case, we're showing the user a static piece of information they will very likely need to use when configuring their OIDC Provider.  Because we're providing the information to the user rather than requesting it from the user, the use of a form field didn't feel right.  My anticipation was that the wording around the URL might be built up a bit more during a user experience review, or at least expanded upon in some end-user documentation.  I further located it in the "Client" section of the page because it just didn't look right separated on the top or of the page, and relates to the client setup.

With that said, I'm really not strongly tied to either approach.  Is there a UX guide or some similar convention documented which would point to one approach v/s the other?

Reply:
> I initially went that way, but changed my mind. :) My reasoning for changing to a hidden field and showing it as plain text was that a disabled text field conveys the impression that the data might vary, but isn't directly mutable by the user. In this case, we're showing the user a static piece of information they will very likely need to use when configuring their OIDC Provider. Because we're providing the information to the user rather than requesting it from the user, the use of a form field didn't feel right. My anticipation was that the wording around the URL might be built up a bit more during a user experience review, or at least expanded upon in some end-user documentation. I further located it in the "Client" section of the page because it just didn't look right separated on the top or of the page, and relates to the client setup.

Fair enough. Could you add `span.show` to the text instead of using `form-text` and remove both `br` then? It will make things cleaner.

> With that said, I'm really not strongly tied to either approach. Is there a UX guide or some similar convention documented which would point to one approach v/s the other?

Not really.

One thing I noticed was that the info is only available in the new/edit pages. It's not displayed in the details page, for example. Would it be relevant to also have it there?

Reply:
Converted to `span.show`.  That's cleaner. :)

I'm not sure about showing the callback URL on the details page; we don't show any of the other client details there, and if there's a client defined, the user should have already set up their provider.  It's really only useful before the end user adds a new connector, since they'll need that when setting up their provider.  The only reason it shows up on the edit page is that It didn't seem worth the work to hide it in the shared template, and it's conceivable that the reminder would be useful at that point.

I could see adding it on the connector `index` page, since the user will be there before they click on "add".  But I'm not really sure _where_ would be a good place to put that information on that page.  Maybe if there was a help link which provided a more prosaic explanation of what this connector is for, it would go in there.  Since we don't have that help function and the info should only be needed at the point where "new" is happening, I'm thinking that having it only on new/edit is probably sufficient.

All that said, I'm not adamantly against having the info elsewhere, so I can stick it wherever if someone has a good idea for how to present it. :)